### PR TITLE
[runtime] Move object manipulation from native to managed.

### PR DESCRIPTION
**Pull request by kumpera**

### CODE REVIEW

**Comment by kumpera** [comment]
@marek-safar please review
@vargaz @lambdageek @ludovic-henry this is one of the strategies we'll have to adopt

File: mcs/class/corlib/System.Globalization/CultureInfo.cs [hunk 1]
```diff
@@ -140,7 +140,15 @@ internal static CultureInfo ConstructCurrentCulture ()
 			if (default_current_culture != null)
 				return default_current_culture;
 
-			var locale_name = get_current_locale_name ();
+			IntPtr locale_name_cstr = default (IntPtr);
+			string locale_name;
```
Comment (vargaz) [inline mcs/class/corlib/System.Globalization/CultureInfo.cs]: Shouldn't this be IntPtr.Zero ?

File: mcs/class/corlib/System.Globalization/CultureInfo.cs [hunk 2]
```diff
@@ -612,7 +620,7 @@ void Construct ()
 		private extern bool construct_internal_locale_from_name (string name);
 
 		[MethodImplAttribute (MethodImplOptions.InternalCall)]
-		private extern static string get_current_locale_name ();
+		private extern static IntPtr get_current_locale_name ();
```
Comment (vargaz) [inline mcs/class/corlib/System.Globalization/CultureInfo.cs]: This needs a corlib version bump.

File: mcs/class/corlib/System.Globalization/CultureInfo.cs [hunk 3]
```diff
@@ -140,7 +140,15 @@ internal static CultureInfo ConstructCurrentCulture ()
...
 
-			var locale_name = get_current_locale_name ();
+			IntPtr locale_name_cstr = default (IntPtr);
+			string locale_name;
+			try {
+				locale_name_cstr = get_current_locale_name ();
+				locale_name = Marshal.PtrToStringAnsi (locale_name_cstr);
+			} finally {
+				Marshal.FreeHGlobal (locale_name_cstr); //FIXME this is wrong on windows.
+			}
```
Comment (vargaz) [inline mcs/class/corlib/System.Globalization/CultureInfo.cs]: This needs to be fixed.

File: mcs/class/corlib/System.Globalization/CultureInfo.cs (hunk 1 above)
Comment (kumpera) [inline mcs/class/corlib/System.Globalization/CultureInfo.cs]: That would be a question of style. I'll change.

File: mcs/class/corlib/System.Globalization/CultureInfo.cs (hunk 3 above)
Comment (kumpera) [inline mcs/class/corlib/System.Globalization/CultureInfo.cs]: With what, do we have an icall that already maps to g_free? All Marshal methods map to all the weird windows alloc functions but not malloc's

File: mcs/class/corlib/System.Globalization/CultureInfo.cs [hunk 4]
```diff
@@ -93,6 +94,28 @@ struct Data {
 			public byte list_sep;
 		}
 
+		/* Native struct is culture-info.h : CultureInfoEntry */
+		const int NUM_CALENDARS = 4;
+		unsafe struct NativeCultureInfo {
```
Comment (marek-safar) [inline mcs/class/corlib/System.Globalization/CultureInfo.cs]: Please move this to separate class/struct (with icalls). That'd help me to migrate to RS CultureInfo (which is still on my todo list). The goal is to have the data part in CultureData class.

**Comment by pr0vieh** [comment]
is this PR related to this https://bugzilla.xamarin.com/show_bug.cgi?id=44795 ?
if it is so any news about this PR ?

**Pull request closed without merging**